File: src-tauri/src/download_engine.rs
use crate::md5_engine;
use crate::s3_client::{self, S3FileInfo};
use crate::state::{DownloadState, FileState, FileStatus};
use aws_sdk_s3::Client;
use md5::{Md5, Digest};
use std::collections::HashMap;
use std::path::Path;
use std::sync::Arc;
use tauri::{AppHandle, Emitter};
use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio::sync::Semaphore;

#[derive(Debug, Clone, serde::Serialize)]
pub struct ProgressEvent {
    pub file_key: String,
    pub downloaded: u64,
    pub total: u64,
    pub speed_bps: u64,
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct OverallProgress {
    pub total_files: usize,
    pub completed_files: usize,
    pub total_bytes: u64,
    pub downloaded_bytes: u64,
    pub speed_bps: u64,
    pub phase: String, // "listing", "downloading", "verifying", "done"
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct VerifyResult {
    pub file_key: String,
    pub status: String, // "match", "mismatch", "no_md5"
    pub expected: Option<String>,
    pub calculated: Option<String>,
}

pub async fn run_download(
    app: AppHandle,
    client: Client,
    bucket: String,
    project: String,
    save_path: String,
    concurrency: usize,
) -> Result<Vec<VerifyResult>, String> {
    // 1. List files
    app.emit("overall-progress", OverallProgress {
        total_files: 0, completed_files: 0,
        total_bytes: 0, downloaded_bytes: 0,
        speed_bps: 0, phase: "listing".to_string(),
    }).ok();

    let all_files = s3_client::list_project_files(&client, &bucket, &project).await?;

    let md5_files: Vec<&S3FileInfo> = all_files.iter().filter(|f| f.is_md5_file).collect();
    let data_files: Vec<&S3FileInfo> = all_files.iter().filter(|f| !f.is_md5_file).collect();

    // 2. Download and parse MD5 files
    let mut md5_map: HashMap<String, String> = HashMap::new();
    for mf in &md5_files {
        let resp = client.get_object()
            .bucket(&bucket)
            .key(&mf.key)
            .send().await
            .map_err(|e| format!("Failed to download MD5 file {}: {}", mf.key, e))?;
        let bytes = resp.body.collect().await
            .map_err(|e| format!("Failed to read MD5 file: {}", e))?
            .into_bytes();
        let content = String::from_utf8_lossy(&bytes);
        md5_map.extend(md5_engine::parse_md5_content(&content));
    }

    // 3. Initialize or resume state
    let mut state = DownloadState::load(&project)
        .unwrap_or(None)
        .filter(|s| s.save_path == save_path)
        .unwrap_or_else(|| DownloadState::new(&project, &bucket, "", &save_path));

    // Compute prefix for stripping to get relative paths
    let prefix = if project.ends_with('/') {
        project.clone()
    } else {
        format!("{}/", project)
    };

    // Create save directory
    std::fs::create_dir_all(&save_path)
        .map_err(|e| format!("Cannot create directory: {}", e))?;

    for f in &data_files {
        // Use relative path (strip project prefix) to preserve directory structure
        let rel_path = f.key.strip_prefix(&prefix).unwrap_or(&f.key).to_string();
        let basename = f.key.rsplit('/').next().unwrap_or(&f.key);
        if !state.files.contains_key(&rel_path) ||
            state.files[&rel_path].status == FileStatus::Failed {
            state.files.insert(rel_path, FileState {
                size: f.size,
                downloaded: 0,
                md5_expected: md5_map.get(basename).cloned(),
                md5_calculated: None,
                status: FileStatus::Pending,
            });
        }
    }
    state.save()?;

    // Emit correct totals before downloads begin
    let total_data_bytes: u64 = data_files.iter().map(|f| f.size).sum();
    app.emit("overall-progress", OverallProgress {
        total_files: data_files.len(),
        completed_files: 0,
        total_bytes: total_data_bytes,
        downloaded_bytes: 0,
        speed_bps: 0,
        phase: "downloading".to_string(),
    }).ok();

    // 4. Download files concurrently
    let semaphore = Arc::new(Semaphore::new(concurrency));
    let app_handle = Arc::new(app.clone());
    let client = Arc::new(client);
    let state_arc = Arc::new(tokio::sync::Mutex::new(state));

    let mut handles = vec![];

    for f in &data_files {
        let rel_path = f.key.strip_prefix(&prefix).unwrap_or(&f.key).to_string();
        let key = f.key.clone();
        let size = f.size;

        // Skip already verified/downloaded files
        {
            let st = state_arc.lock().await;
            if let Some(fs) = st.files.get(&rel_path) {
                if fs.status == FileStatus::Verified || fs.status == FileStatus::Downloaded {
                    continue;
                }
            }
        }

        let sem = semaphore.clone();
        let app_h = app_handle.clone();
        let client_c = client.clone();
        let bucket_c = bucket.clone();
        let save_c = save_path.clone();
        let state_c = state_arc.clone();

        let handle = tokio::spawn(async move {
            let _permit = sem.acquire().await.map_err(|e| e.to_string())?;

            // Update status
            {
                let mut st = state_c.lock().await;
                if let Some(fs) = st.files.get_mut(&rel_path) {
                    fs.status = FileStatus::Downloading;
                }
            }

            // Preserve directory structure: join relative path with save dir
            let file_path = Path::new(&save_c).join(&rel_path);
            // Create parent directories if needed
            if let Some(parent) = file_path.parent() {
                tokio::fs::create_dir_all(parent).await
                    .map_err(|e| format!("Cannot create directory {}: {}", parent.display(), e))?;
            }

            // Check for resume: if file exists partially, use Range header
            let existing_size = tokio::fs::metadata(&file_path).await
                .map(|m| m.len()).unwrap_or(0);

            let mut req = client_c.get_object().bucket(&bucket_c).key(&key);
            let start_byte = if existing_size > 0 && existing_size < size {
                req = req.range(format!("bytes={}-", existing_size));
                existing_size
            } else {
                0u64
            };

            let resp = req.send().await
                .map_err(|e| format!("S3 download error for {}: {}", rel_path, e))?;

            // Open file for writing (append if resuming, create if new)
            let mut file = if start_byte > 0 {
                tokio::fs::OpenOptions::new()
                    .append(true).open(&file_path).await
                    .map_err(|e| format!("Cannot open file for append: {}", e))?
            } else {
                tokio::fs::File::create(&file_path).await
                    .map_err(|e| format!("Cannot create file {}: {}", rel_path, e))?
            };

            let mut reader = resp.body.into_async_read();
            let mut hasher = Md5::new();
            let mut downloaded = start_byte;
            let start_time = std::time::Instant::now();

            // If resuming, we need to hash the existing part first
            if start_byte > 0 {
                let existing_data = tokio::fs::read(&file_path).await
                    .map_err(|e| format!("Cannot read existing file: {}", e))?;
                hasher.update(&existing_data[..start_byte as usize]);
            }

            let mut buf = vec![0u8; 256 * 1024]; // 256KB chunks
            loop {
                let n = reader.read(&mut buf).await
                    .map_err(|e| format!("Stream error: {}", e))?;
                if n == 0 {
                    break;
                }
                let chunk = &buf[..n];
                file.write_all(chunk).await
                    .map_err(|e| format!("Write error: {}", e))?;
                hasher.update(chunk);
                downloaded += n as u64;

                let elapsed = start_time.elapsed().as_secs_f64();
                let speed = if elapsed > 0.0 {
                    ((downloaded - start_byte) as f64 / elapsed) as u64
                } else { 0 };

                app_h.emit("file-progress", ProgressEvent {
                    file_key: rel_path.clone(),
                    downloaded,
                    total: size,
                    speed_bps: speed,
                }).ok();

                // Periodically save state
                if downloaded % (10 * 1024 * 1024) < n as u64 {
                    let mut st = state_c.lock().await;
                    if let Some(fs) = st.files.get_mut(&rel_path) {
                        fs.downloaded = downloaded;
                    }
                    st.save().ok();
                }
            }

            file.flush().await.map_err(|e| format!("Flush error: {}", e))?;

            let md5_hex = format!("{:x}", hasher.finalize());

            // Update state
            {
                let mut st = state_c.lock().await;
                if let Some(fs) = st.files.get_mut(&rel_path) {
                    fs.downloaded = downloaded;
                    fs.md5_calculated = Some(md5_hex.clone());
                    fs.status = FileStatus::Downloaded;
                }
                st.save().ok();
            }

            Ok::<(String, String), String>((rel_path, md5_hex))
        });

        handles.push(handle);
    }

    // 5. Collect results and verify
    let mut results = Vec::new();

    for handle in handles {
        match handle.await {
            Ok(Ok((rel_path, md5_calc))) => {
                let mut st = state_arc.lock().await;
                let file_state = st.files.get_mut(&rel_path);

                let (status, expected) = if let Some(fs) = file_state {
                    if let Some(ref expected) = fs.md5_expected {
                        if expected == &md5_calc {
                            fs.status = FileStatus::Verified;
                            ("match".to_string(), Some(expected.clone()))
                        } else {
                            fs.status = FileStatus::Failed;
                            ("mismatch".to_string(), Some(expected.clone()))
                        }
                    } else {
                        fs.status = FileStatus::Downloaded;
                        ("no_md5".to_string(), None)
                    }
                } else {
                    ("no_md5".to_string(), None)
                };

                st.save().ok();
                results.push(VerifyResult {
                    file_key: rel_path,
                    status,
                    expected,
                    calculated: Some(md5_calc),
                });
            }
            Ok(Err(e)) => {
                results.push(VerifyResult {
                    file_key: "unknown".to_string(),
                    status: format!("error: {}", e),
                    expected: None,
                    calculated: None,
                });
            }
            Err(e) => {
                results.push(VerifyResult {
                    file_key: "unknown".to_string(),
                    status: format!("task error: {}", e),
                    expected: None,
                    calculated: None,
                });
            }
        }
    }

    app.emit("overall-progress", OverallProgress {
        total_files: data_files.len(),
        completed_files: results.len(),
        total_bytes: data_files.iter().map(|f| f.size).sum(),
        downloaded_bytes: data_files.iter().map(|f| f.size).sum(),
        speed_bps: 0,
        phase: "done".to_string(),
    }).ok();

    Ok(results)
}

File: src/downloader/lib/download-engine.ts
import SparkMD5 from "spark-md5";
import type { S3Client } from "@aws-sdk/client-s3";
import { getObjectStream, getObjectText, type S3FileInfo } from "./s3-browser";
import { parseMd5Content } from "./md5-utils";
import type { VerifyResult, FileProgress, OverallProgress } from "./types";

interface DownloadOptions {
  client: S3Client;
  bucket: string;
  project: string;
  files: S3FileInfo[];
  dirHandle: FileSystemDirectoryHandle;
  concurrency?: number;
  onFileProgress?: (progress: FileProgress) => void;
  onOverallProgress?: (progress: OverallProgress) => void;
}

// Flush interval: close and reopen writable to commit partial data to disk.
// FileSystemWritableFileStream uses a swap file — data is only persisted on close().
// Without periodic flushing, all progress is lost if the page closes mid-download.
const FLUSH_BYTES = 50 * 1024 * 1024; // 50MB

class Semaphore {
  private queue: (() => void)[] = [];
  private running = 0;

  constructor(private max: number) {}

  async acquire(): Promise<void> {
    if (this.running < this.max) {
      this.running++;
      return;
    }
    return new Promise((resolve) => {
      this.queue.push(() => {
        this.running++;
        resolve();
      });
    });
  }

  release(): void {
    this.running--;
    const next = this.queue.shift();
    if (next) next();
  }
}

async function getOrCreateSubDir(
  root: FileSystemDirectoryHandle,
  path: string,
): Promise<FileSystemDirectoryHandle> {
  const parts = path.split("/").filter(Boolean);
  let current = root;
  for (const part of parts) {
    current = await current.getDirectoryHandle(part, { create: true });
  }
  return current;
}

export async function runDownload(
  options: DownloadOptions,
): Promise<VerifyResult[]> {
  const {
    client,
    bucket,
    project,
    files,
    dirHandle,
    concurrency = 3,
    onFileProgress,
    onOverallProgress,
  } = options;

  const prefix = project
    ? project.endsWith("/")
      ? project
      : `${project}/`
    : "";

  // Phase: listing
  onOverallProgress?.({
    totalFiles: 0,
    completedFiles: 0,
    totalBytes: 0,
    downloadedBytes: 0,
    speedBps: 0,
    phase: "listing",
  });

  // Separate MD5 and data files
  const md5Files = files.filter((f) => f.isMd5File);
  const dataFiles = files.filter((f) => !f.isMd5File);

  // Download and parse all MD5 files
  const md5Map = new Map<string, string>();
  for (const mf of md5Files) {
    const content = await getObjectText(client, bucket, mf.key);
    for (const [name, hash] of parseMd5Content(content)) {
      md5Map.set(name, hash);
    }
  }

  const totalBytes = dataFiles.reduce((sum, f) => sum + f.size, 0);

  // Phase: downloading
  onOverallProgress?.({
    totalFiles: dataFiles.length,
    completedFiles: 0,
    totalBytes,
    downloadedBytes: 0,
    speedBps: 0,
    phase: "downloading",
  });

  // Track overall progress
  let completedFiles = 0;
  let totalDownloaded = 0;
  const overallStart = performance.now();

  const emitOverall = () => {
    const elapsed = (performance.now() - overallStart) / 1000;
    onOverallProgress?.({
      totalFiles: dataFiles.length,
      completedFiles,
      totalBytes,
      downloadedBytes: totalDownloaded,
      speedBps: elapsed > 0 ? Math.round(totalDownloaded / elapsed) : 0,
      phase: "downloading",
    });
  };

  const semaphore = new Semaphore(concurrency);
  const results: VerifyResult[] = [];

  const downloadFile = async (file: S3FileInfo): Promise<VerifyResult> => {
    await semaphore.acquire();
    try {
      const relPath = file.key.startsWith(prefix)
        ? file.key.slice(prefix.length)
        : file.key;
      const basename = relPath.split("/").pop() ?? relPath;

      // Create subdirectories if needed
      const dirParts = relPath.split("/");
      dirParts.pop(); // remove filename
      const parentDir =
        dirParts.length > 0
          ? await getOrCreateSubDir(dirHandle, dirParts.join("/"))
          : dirHandle;

      // Get or create file handle
      const fileHandle = await parentDir.getFileHandle(basename, {
        create: true,
      });

      // Check existing file for resume
      const existingFile = await fileHandle.getFile();
      const existingSize = existingFile.size;
      const isComplete = existingSize === file.size && existingSize > 0;
      const isPartial = existingSize > 0 && existingSize < file.size;

      console.log(
        `[download] ${relPath}: local=${existingSize}, remote=${file.size}` +
          (isComplete ? " → SKIP (complete)" : isPartial ? " → RESUME" : " → NEW"),
      );

      const spark = new SparkMD5.ArrayBuffer();
      let downloaded = 0;
      const fileStart = performance.now();

      // Read existing bytes into MD5 hasher
      if ((isComplete || isPartial) && existingSize > 0) {
        const existReader = existingFile.stream().getReader();
        while (true) {
          const { done, value } = await existReader.read();
          if (done) break;
          spark.append(value.buffer as ArrayBuffer);
        }
        downloaded = existingSize;
        totalDownloaded += existingSize;

        onFileProgress?.({
          fileKey: relPath,
          downloaded,
          total: file.size,
          speedBps: 0,
        });
        emitOverall();
      }

      // Download remaining bytes if needed
      if (!isComplete) {
        let writable = isPartial
          ? await fileHandle.createWritable({ keepExistingData: true })
          : await fileHandle.createWritable();

        if (isPartial) {
          await writable.seek(existingSize);
        }

        const stream = await getObjectStream(
          client,
          bucket,
          file.key,
          isPartial ? existingSize : undefined,
        );
        const reader = stream.getReader();
        let bytesSinceFlush = 0;

        while (true) {
          const { done, value } = await reader.read();
          if (done) break;

          await writable.write(value as unknown as BufferSource);
          spark.append(value.buffer as ArrayBuffer);
          downloaded += value.byteLength;
          totalDownloaded += value.byteLength;
          bytesSinceFlush += value.byteLength;

          const elapsed = (performance.now() - fileStart) / 1000;
          const speed = elapsed > 0 ? downloaded / elapsed : 0;

          onFileProgress?.({
            fileKey: relPath,
            downloaded,
            total: file.size,
            speedBps: Math.round(speed),
          });
          emitOverall();

          // Periodic flush: commit partial data to disk so resume works
          // after page close. close() commits the swap file, then we reopen.
          if (bytesSinceFlush >= FLUSH_BYTES) {
            await writable.close();
            writable = await fileHandle.createWritable({ keepExistingData: true });
            await writable.seek(downloaded);
            bytesSinceFlush = 0;
          }
        }

        await writable.close();
      }

      completedFiles++;

      const md5Calc = spark.end();
      const expectedMd5 = md5Map.get(basename) ?? null;

      let status: VerifyResult["status"];
      if (!expectedMd5) {
        status = "no_md5";
      } else if (expectedMd5 === md5Calc) {
        status = "match";
      } else {
        status = "mismatch";
      }

      return {
        fileKey: relPath,
        status,
        expected: expectedMd5,
        calculated: md5Calc,
      };
    } catch (e) {
      console.error(`[download] error for ${file.key}:`, e);
      return {
        fileKey: file.key,
        status: "error",
        expected: null,
        calculated: null,
      };
    } finally {
      semaphore.release();
    }
  };

  // Launch all downloads (semaphore limits concurrency)
  const promises = dataFiles.map((f) => downloadFile(f));
  const settled = await Promise.allSettled(promises);

  for (const result of settled) {
    if (result.status === "fulfilled") {
      results.push(result.value);
    } else {
      results.push({
        fileKey: "unknown",
        status: "error",
        expected: null,
        calculated: null,
      });
    }
  }

  // Phase: done
  onOverallProgress?.({
    totalFiles: dataFiles.length,
    completedFiles: dataFiles.length,
    totalBytes,
    downloadedBytes: totalBytes,
    speedBps: 0,
    phase: "done",
  });

  return results;
}
